File: app/src/main/java/com/lintaostudy/kotlin/demo/java/SingleTon1.java
package com.lintaostudy.kotlin.demo.java;

/**
 * @author :created by lintao
 * 版本：1.0
 * 创建日期：2021/5/11 20
 * 描述： 静态内部类单例模式
 */
public class SingleTon1 {

    private static class SingleTonProvider {

        private static SingleTon1 instance = new SingleTon1();
    }

    private SingleTon1() {
    }

    public static SingleTon1 getInstance() {
        return SingleTonProvider.instance;
    }
}

File: app/src/main/java/com/lintaostudy/demo/kotlin/SingleTon1.kt
package com.lintaostudy.demo.kotlin

/**
 * @author :created by lintao
 * 版本：1.0
 * 创建日期：2021/5/11 20
 * 描述： kotlin 实现静态内部类单例模式
 * 推荐使用静态内部类单例模式:
 * 1. 既能保证线程安全
 * 2. 单例延迟实例化(在加载外部类的时候不会加载内部类,调用SingleProvider.holder才会加载)
 * 3. 保证对象的唯一性
 */
class SingleTon1 private constructor() {
    companion object {
        val instance = SingleProvider.holder
    }

    private object SingleProvider {
        val holder = SingleTon1
    }
}